modules/kafka/tests/test_kafka_topic.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2020 Balabit
+ * Copyright (c) 2020 Balazs Scheidler
+ * Copyright (c) 2020 Vivin Peris
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include <syslog-ng.h>
+#include <apphook.h>
+#include "logthrdest/logthrdestdrv.h"
+#include "kafka-dest-driver.h"
+#include <librdkafka/rdkafka.h>
+#include "kafka-dest-worker.h"
+#include "kafka-dest-driver.h"
+#include "kafka-dest-worker-internal.h"
+
+#include <criterion/criterion.h>
+#include <criterion/parameterized.h>
+
+struct topic_validity_test_params
+{
+  gchar *topic_name;
+  gboolean isvalid;
+};
+
+static void
+setup(void)
+{
+  app_startup();
+}
+
+static void
+teardown(void)
+{
+  app_shutdown();
+}
+
+TestSuite(kafka_topic, .init = setup, .fini = teardown);
+
+ParameterizedTestParameters(kafka_topic, topic_validity_tests)
+{
+  static struct topic_validity_test_params params[] =
+  {
+    {"test 1", FALSE},
+    {"", FALSE},
+    {"..", FALSE},
+    {".", FALSE},
+    {" invalidname", FALSE},
+    {"a*123", FALSE},
+    {"topic@123", FALSE},
+    {"aba#2", FALSE},
+    {"dontuse&", FALSE},
+    {"illegal%", FALSE},
+    {"name!", FALSE},
+    {"topics^", FALSE},
+    {"round()", FALSE},
+    {"sum+", FALSE},
+    {"topic=name", FALSE},
+    {"topic``~invalid", FALSE},
+    {STRING_250_LEN, FALSE},
+    {"validname123", TRUE},
+    {"valid.name123", TRUE},
+    {"valid-name123", TRUE},
+    {"valid_name123", TRUE},
+    {"valid.name-123", TRUE},
+    {"valid_name-123", TRUE},
+    {"This.is.a.valid.name", TRUE},
+    {"This.is-a_valid-name", TRUE},
+    {STRING_249_LEN, TRUE},
+  };
+
+  return cr_make_param_array(struct topic_validity_test_params, params, sizeof(params)/sizeof(params[0]));
+}
+
+ParameterizedTest(struct topic_validity_test_params *param, kafka_topic, topic_validity_tests)
+{
+  KafkaDestDriver *driver = (KafkaDestDriver *) kafka_dd_new(configuration);
+  cr_assert_eq(kafka_dd_validate_topic_name(driver, param->topic_name), param->isvalid);
```

Comment:
just a nitpick, but you don't need to change it: if the kafka_dd_validate_topic_name() would return the error in a GError object, https://developer.gnome.org/glib/stable/glib-Error-Reporting.html, then you could move all the error messages currently inside this function to its caller. And with that, you could completely remove the "self" argument, making it
static (as C++ calls this).

With it static, this testcase would be much simpler, you wouldn't need to allocate/free an entire KafkaDestDriver instance just to test a string validation function.

Reply:
I really like this idea.